WIP experiments for persistence API for typed

raboof opened the pull request:
Early straw man idea sketching

raboof committed:
Some experiments with a typed api for persistence

raboof committed:
Expand example with a timer

raboof committed:
Move 'onRecoveryComplete' to 'builder' style

rather than using a default parameter which is hard to
evolve with binary compatibility

raboof committed:
Explicit snapshots

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,239 @@
+        onState: State ⇒ Option[Snapshot]          = (_: State) ⇒ None,
+        onEvent: Event ⇒ Option[Snapshot]          = (_: Event) ⇒ None,
+        on:      (State, Event) ⇒ Option[Snapshot] = (_: State, _: Event) ⇒ None,
+        recover: Snapshot ⇒ Option[State]
```
what is `Snapshot` vs. `State`? isn't the type of the stored snapshot always `State`?

raboof replied:
My thinking here was that `State` might be (entirely) serializable: for example the `ExplicitSnapshots` example below only takes snapshot when there are not tasks in flight, meaning tasks don't need to be serializable.

That example is a bit contrived because the tasks are serialized in the events anyway, but in general I've found it to be helpful to separate the events and snapshots (which are persistent and can be tricky to evolve) from the commands and state (which are transient and can therefore be refactored freely).

patriknw replied:
I find that complicated. Shouldn't the serializer know how to ignore transient fields if any.

raboof replied:
That might be dangerous though: developers might expect certain fields to survive serialization while they don't. I like how refactoring `State` for 'logic-only' changes would only affect the conversions in `onState`/`recover`, and not the serialization itself. Maybe it'd be good to get some data on how snapshotting is used in practice?

patriknw replied:
The snapshot must anyway be a subset of the state and it must be possible to create the state from the snapshot. That only leaves room for transient fields (e.g. a cache), and if you don't want to handle that in the serializer I think it's more like `state.clearForSnapshot()` before you save it. And it must anyway immutable.

I see your point in having different classes because you want to evolve them differently, i.e. this is a mapping layer. Not sure we should have yet another way of mapping things for evolving. We already have guidelines for how to do that with serialization (and event adapaters).

If going down this path I think it should be a pure mapping function `Snapshot => State`, `State => Snapshot` instead of convoluting it with **when** the snapshot is to be taken.

patriknw replied:
Another parameter, I think we should also support automatic snapshotting after a given number of events. That should be the normal use of snapshots, a plain optimization.

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,239 @@
+      def snapshot[Snapshot](
+        onState: State ⇒ Option[Snapshot]          = (_: State) ⇒ None,
+        onEvent: Event ⇒ Option[Snapshot]          = (_: Event) ⇒ None,
+        on:      (State, Event) ⇒ Option[Snapshot] = (_: State, _: Event) ⇒ None,
```
another way would be to have a `PersistentActorContext` with a `snapshot()` method that can be called

raboof committed:
Example spawning a child actor

raboof committed:
Taking a snapshot needs access to the state

raboof committed:
Added signal handler

I like how simple the types of `commandHandler` and
`signalHandler` are now, but it'd be nice if they were closer together and could perhaps even share the `state` switch. I'll give that a try.

raboof committed:
Attempt to get signal handling closer to command handling

raboof commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,317 @@
+    class ActionHandler[Command: ClassTag, Event, State](val handler: ((ActorContext[Command], State, Any) ⇒ PersistentEffect[Event])) {
+      def onSignal(signalHandler: (ActorContext[Command], State, Any) ⇒ PersistentEffect[Event]): ActionHandler[Command, Event, State] =
+        new ActionHandler({
+          case (ctx, state, command: Command) ⇒ handler(ctx, state, command)
```
This implies signals cannot also be commands - which seems reasonable, would be nice to ensure this statically...

raboof committed:
Always use the full state for snapshots

If you want to add a mapping layer use event adapters or a custom
serializer.

raboof committed:
Example of an actor that separates 'persistent' and transient state

raboof committed:
Use ordering 'command, event, state, context' consistently

raboof committed:
Allow multiple `andThen`

raboof committed:
Replace ask with spawned adapters

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,394 @@
+
+      commandHandler = ActionHandler.cmd({
+        case Cmd(data) ⇒ Persist(Evt(data))
+      }),
```
So `.cmd` is a convenience for the case when state and context is not needed. Not sure it's worth having another way of doing that, since if you later need them for one command you need to refactor.
Doesn't save much compared to 
```ActionHandler  {(cmd, _, _) => cmd match {```

By the way, `({` is not needed, this reads better (in my opinion):
```
    ActionHandler.cmd {
        case Cmd(data) ⇒ Persist(Evt(data))
    }

raboof replied:
:+1: on the `({`. `.cmd` in indeed just convenience to save a `{(cmd, _, _) => cmd match {` which looks a bit complicated/noisy for that case imo...

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,394 @@
+      commandHandler = ActionHandler.cmd({
+        case Cmd(data, sender) ⇒
+          Persist(Evt(data))
+            .andThen { evt ⇒ { sender ! Ack } }
```
The parameter to `andThen` should be the new state, not the event. The event is already known.

patriknw replied:
we haven't solved how it can stop itself, perhaps it could be `.andThenStop()`

raboof replied:
We could also promote `.andThen { _ => ctx.stop(ctx.self) }`?

raboof replied:
Hmm, you're absolutely right that `andThen` should pass in the updated `state`. Actually kind of dangerous that we so easily close over the 'old' state there... but I guess we can't really avoid that.

patriknw replied:
`ctx.stop` is for children, does it even work for self?

raboof replied:
[afaik yes](http://doc.akka.io/docs/akka/current/scala/actors.html#stopping-actors) :)

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,394 @@
+            case (correlationId, data) ⇒ performSideEffect(ctx.self, correlationId, data)
+          }
+        }
+      }
```
Perhaps write one of the more complex examples with by breaking out the different functions in separate vals. In my taste this becomes too many things stacked after each other (different params, and various fluent api calls) and it's hard to see where one ends and another begins.

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,394 @@
+      def byState[Command: ClassTag, Event, State](actionHandler: State ⇒ ActionHandler[Command, Event, State]): ActionHandler[Command, Event, State] =
+        new ActionHandler(handler = {
+          case (action, state, ctx) ⇒ actionHandler(state).handler(action, state, ctx)
+        })
```
That is an interesting alternative to the `State => Actions`, what are the pros and cons?
Perhaps this is better because it's more explicit in the types: `commandHandler: ActionHandler`.

patriknw commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,394 @@
+    case class Unhandled[Event]() extends PersistentEffect[Event]
+
+    class ActionHandler[Command: ClassTag, Event, State](val handler: ((Any, State, ActorContext[Command]) ⇒ PersistentEffect[Event])) {
+      def onSignal(signalHandler: (Any, State, ActorContext[Command]) ⇒ PersistentEffect[Event]): ActionHandler[Command, Event, State] =
```
shouldn't this be a partial function, we have that for `Actor.immutable` ?

patriknw reviewed (commented):
This an excellent API prototype, I like it.

ktoso commented on akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala:
```diff
@@ -0,0 +1,394 @@
+
+    case class Persist[Event](event: Event, callbacks: List[Event ⇒ Unit] = Nil) extends PersistentEffect[Event] {
+      def andThen(callback: Event ⇒ Unit) = Persist(event, callback :: callbacks)
+    }
```
Same as I mentioned on the multi-dc API: https://github.com/typesafehub/akka-commercial-addons/pull/107#discussion_r138634348

All things must be able to andThen otherwise we make it hard for people to abstract "and do this thing always anyway".

raboof replied:
Hmm, I generally agree, though this did get kind of hairy (https://github.com/akka/akka/pull/23631/commits/ad11555c0d6c21ec36e9b2f6fe5f45521b20f1dc) - do you have a nicer pattern? I tried moving `callbacks` to `PersistentEffect` but that caused variance problems I wasn't sure how to get rid of.

patriknw replied:
isn't that just implementation details, I solved it at the other place

raboof replied:
Yes, that's an implementation thing. Cleaned it up a bit, still not entirely happy with it - having everything inside the `ReplicatedEntity` abstract superclass does make the typing/generics a bit nicer, but that wouldn't be very 'akka-typed-like'...

raboof committed:
Allow 'andThen' on each PersistentEffect

raboof committed:
Make 'onSignal' a partial function

raboof committed:
Remove unnecessary parens

raboof committed:
Simplify 'andThen', make sure it receives the State

patriknw commented:
Superseded by https://github.com/akka/akka/pull/23674, which also includes the excellent API examples.
